src/app/src/main/kotlin/org/imperial/mrc/hint/exceptions/HintExceptionHandler.kt

```diff
@@ -3,41 +3,44 @@ package org.imperial.mrc.hint.exceptions
 import org.imperial.mrc.hint.AppProperties
 import org.imperial.mrc.hint.models.ErrorDetail
 import org.imperial.mrc.hint.models.ErrorDetail.Companion.defaultError
+import org.springframework.beans.TypeMismatchException
 import org.springframework.core.Ordered
 import org.springframework.core.annotation.Order
-import org.springframework.http.HttpHeaders
 import org.springframework.http.HttpStatus
 import org.springframework.http.ResponseEntity
-import org.springframework.lang.Nullable
+import org.springframework.http.converter.HttpMessageNotReadableException
+import org.springframework.web.HttpMediaTypeNotAcceptableException
+import org.springframework.web.HttpMediaTypeNotSupportedException
+import org.springframework.web.HttpRequestMethodNotSupportedException
+import org.springframework.web.bind.MethodArgumentNotValidException
+import org.springframework.web.bind.MissingServletRequestParameterException
+import org.springframework.web.bind.ServletRequestBindingException
 import org.springframework.web.bind.annotation.ControllerAdvice
 import org.springframework.web.bind.annotation.ExceptionHandler
 import org.springframework.web.context.request.WebRequest
-import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
+import org.springframework.web.context.request.async.AsyncRequestTimeoutException
+import org.springframework.web.multipart.support.MissingServletRequestPartException
+import org.springframework.web.servlet.ModelAndView
+import org.springframework.web.servlet.NoHandlerFoundException
 import java.lang.reflect.UndeclaredThrowableException
+import java.net.BindException
 import java.text.MessageFormat
 import java.util.*
 
 @Order(Ordered.HIGHEST_PRECEDENCE)
 @ControllerAdvice
 class HintExceptionHandler(private val errorCodeGenerator: ErrorCodeGenerator,
                            private val appProperties: AppProperties)
-    : ResponseEntityExceptionHandler()
 {
 
-    override fun handleExceptionInternal(e: Exception,
-                                         @Nullable body: Any?,
-                                         headers: HttpHeaders,
-                                         status: HttpStatus,
-                                         request: WebRequest): ResponseEntity<Any>
+    @ExceptionHandler(NoHandlerFoundException::class)
+    fun handleNoHandlerFoundException(): ModelAndView
     {
-        logger.error(e.message)
-        // this handles standard Spring MVC exceptions which do not contain
-        // sensitive info so we return the original error message here
-        return unexpectedError(status, request, e.message)
+        return viewErrorPage("404")
```

Comment:
This really should distinguish between api requests and web requests - we shouldn't really be getting any 404s over the api calls from the javascript, but if we do it's going confuse things for it to return an html page with a 200 rather than json with a 404. Should be able to tell if this is an api request from the request object. We haven't got a dedicated api route prefix in this application, but we should be including application/json in the Accept header of our api requests so you could check for that.